From 5f6c17dc9f54899cae48c4d6acaa5ddf7a727780 Mon Sep 17 00:00:00 2001
From: Winson Chung <winsonc@google.com>
Date: Mon, 14 Jul 2025 23:35:55 +0000
Subject: [PATCH 02/56] Use provided calling uid/pid for checking
 intercept-global-drag flag

- This particular path can not rely on ActivityTaskManagerService's
  enforceTaskPermission() call since that uses the binder calling
  uid/pid, which has been cleared prior to calling
  validateAddingWindowLw().

Bug: 428700812
Flag: EXEMPT bugfix
Test: atest DragDropControllerTests

(cherry picked from commit 467b18585de4d1faa80d4b056dd3d69654d16651)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:c63342c195420912e6b2ce30b8a13d435f253a05
Merged-In: Ic4de10db891c79130bb6e2f4d6e519ca6e4c0f73
Change-Id: Ic4de10db891c79130bb6e2f4d6e519ca6e4c0f73
---
 services/core/java/com/android/server/wm/DisplayPolicy.java   | 4 +++-
 .../src/com/android/server/wm/DragDropControllerTests.java    | 3 ++-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/services/core/java/com/android/server/wm/DisplayPolicy.java b/services/core/java/com/android/server/wm/DisplayPolicy.java
index 10e393fd5..0ee6214e1 100644
--- a/services/core/java/com/android/server/wm/DisplayPolicy.java
+++ b/services/core/java/com/android/server/wm/DisplayPolicy.java
@@ -1128,7 +1128,9 @@ public class DisplayPolicy {
                     "DisplayPolicy");
         }
         if ((attrs.privateFlags & PRIVATE_FLAG_INTERCEPT_GLOBAL_DRAG_AND_DROP) != 0) {
-            ActivityTaskManagerService.enforceTaskPermission("DisplayPolicy");
+            mContext.enforcePermission(
+                    android.Manifest.permission.MANAGE_ACTIVITY_TASKS, callingPid, callingUid,
+                    "DisplayPolicy");
         }
 
         final String systemUiPermission =
diff --git a/services/tests/wmtests/src/com/android/server/wm/DragDropControllerTests.java b/services/tests/wmtests/src/com/android/server/wm/DragDropControllerTests.java
index 523b72351..37af83ca0 100644
--- a/services/tests/wmtests/src/com/android/server/wm/DragDropControllerTests.java
+++ b/services/tests/wmtests/src/com/android/server/wm/DragDropControllerTests.java
@@ -260,7 +260,8 @@ public class DragDropControllerTests extends WindowTestsBase {
         attrs.privateFlags |= PRIVATE_FLAG_INTERCEPT_GLOBAL_DRAG_AND_DROP;
         policy.validateAddingWindowLw(attrs, Binder.getCallingPid(), Binder.getCallingUid());
 
-        verify(mWm.mAtmService).enforceTaskPermission(any());
+        verify(mWm.mContext).enforcePermission(
+                eq(android.Manifest.permission.MANAGE_ACTIVITY_TASKS), anyInt(), anyInt(), any());
     }
 
     @Test
-- 
2.52.0

